include/hip/hcc_detail/math_functions.h

```diff
@@ -20,237 +20,1125 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+__device__
+inline
+int isinf(float x) { return __ocml_isinf_f32(x); }
+__device__
+inline
+int isnan(float x) { return __ocml_isnan_f32(x); }
+__device__
+inline
+float j0f(float x) { return __ocml_j0_f32(x); }
+__device__
+inline
+float j1f(float x) { return __ocml_j1_f32(x); }
+__device__
+inline
+float jnf(int n, float x)
+{   // TODO: we could use Ahmes multiplication and the Miller & Brown algorithm
+    //       for linear recurrences to get O(log n) steps, but it's unclear if
+    //       it'd be beneficial in this case.
+    if (n == 0) return j0f(x);
+    if (n == 1) return j1f(x);
+
+    float x0 = j0f(x);
+    float x1 = j1f(x);
+    for (int i = 1; i < n; ++i) {
+        float x2 = (2 * i) / x * x1 - x0;
+        x0 = x1;
+        x1 = x2;
+    }
+
+    return x1;
+}
+__device__
+inline
+float ldexpf(float x, int e) { return __ocml_ldexp_f32(x, e); }
+__device__
+inline
+float lgammaf(float x) { return __ocml_lgamma_f32(x); }
+__device__
+inline
+long long int llrintf(float x) { return __ocml_rint_f32(x); }
+__device__
+inline
+long long int llroundf(float x) { return __ocml_round_f32(x); }
+__device__
+inline
+float log10f(float x) { return __ocml_log10_f32(x); }
+__device__
+inline
+float log1pf(float x) { return __ocml_log1p_f32(x); }
+__device__
+inline
+float log2f(float x) { return __ocml_log2_f32(x); }
+__device__
+inline
+float logbf(float x) { return __ocml_logb_f32(x); }
+__device__
+inline
+float logf(float x) { return __ocml_log_f32(x); }
+__device__
+inline
+long int lrintf(float x) { return __ocml_rint_f32(x); }
+__device__
+inline
+long int lroundf(float x) { return __ocml_round_f32(x); }
+__device__
+inline
+float modff(float x, float* iptr)
+{
+    float tmp;
+    float r =
+        __ocml_modf_f32(x, (__attribute__((address_space(5))) float*) &tmp);
+    *iptr = tmp;
+
+    return r;
+}
+__device__
+inline
+float nanf(const char* tagp)
+{
+    union {
+        float val;
+        struct ieee_float {
+            uint32_t mantissa : 22;
+            uint32_t quiet : 1;
+            uint32_t exponent : 8;
+            uint32_t sign : 1;
+        } bits;
+
+        static_assert(sizeof(float) == sizeof(ieee_float), "");
+    } tmp;
+
+    tmp.bits.sign = 0u;
+    tmp.bits.exponent = ~0u;
+    tmp.bits.quiet = 1u;
+    tmp.bits.mantissa = __make_mantissa(tagp);
+
+    return tmp.val;
+}
+__device__
+inline
+float nearbyintf(float x) { return __ocml_nearbyint_f32(x); }
+__device__
+inline
+float nextafterf(float x, float y) { return __ocml_nextafter_f32(x, y); }
+__device__
+inline
+float norm3df(float x, float y, float z) { return __ocml_len3_f32(x, y, z); }
+__device__
+inline
+float norm4df(float x, float y, float z, float w)
+{
+    return __ocml_len4_f32(x, y, z, w);
+}
+__device__
+inline
+float normcdff(float x) { return __ocml_ncdf_f32(x); }
+__device__
+inline
+float normcdfinvf(float x) { return __ocml_ncdfinv_f32(x); }
+__device__
+inline
+float normf(int dim, const float* a)
```

Comment:
Accuracy is poor and no overflow protection.  Better add comment that this is just a placeholder until there is demand for this function.